src/lib.js

```diff
@@ -88,8 +86,42 @@ export const query = (db, { select, where }) => {
     }
   }
 
-  const matches = evaluate(db, {
-    And: clauses.sort(byClause),
+
+  /**
+   * Ferret out all Cases, inclusive of those nested within Not and Rule clauses
+   * to ensure retreival of all related facts.
+   */
+
+  const cases = []
+
+  /** @type {Array<API.Clause>} */
+  const clauseStack = [...clauses]
+
+  while (clauseStack.length) {
+    const clause = clauseStack[0]
+    clauseStack.shift()
+    if (clause.Case) {
+      cases.push(clause.Case)
+    } else if (clause.And) {
+      clauseStack.unshift(...clause.And)
+    } else if (clause.Or) {
+      clauseStack.unshift(...clause.Or)
+    } else if (clause.Not) {
+      clauseStack.unshift(clause.Not)
+    } else if (clause.Rule) {
+      clauseStack.unshift(clause.Rule.rule.where)
+    }
+  }
+
+  // Collect all pertinent facts
+  const facts = await Promise.all(cases.map(c => iterateFacts(db, c)))
```

Comment:
I think this may be a little too naive way to go about it. Consider following example that illustrates the issue

```edn
[?entity :user/name "gozala"]  ;; This will do a good index lookup
[?entity :address ?address]    ;; This will have to get everything that has address and addresses
;; This gets very contrived 
[?entity ?relation "irakli"]   ;; This needs to fetch anything that references "irakli"
[?q ?relation ?address]        ;; This will have to fetch all records
```

My example gets very contrived on line 3, but even on 2nd line we end having to fetch a lot more records than what we had to if we have done it serially. However worst case (which my example isn't illustrating all that well) is on the last line where we'd be fetching all of the records which would make this impractical.